10-JavaScript/classes.md

```diff
@@ -0,0 +1,289 @@
+
+### Constructor Methods
+
+The constructor method is a special method named `constructor` which is called when a new instance is created.  You can think of it like an `initialize` method from Ruby.  Constructors exist to start our instances off in a proper state.  Use them to set your attributes with initial values using `this.<attributeName>` to create attributes (similar to using @ in Ruby).
+
+```javascript
+class Book {
+  constructor(title, author) {
+    console.log("Constructor Ran!");
+    this.title = title;
+    this.author = author;
+  }
+}
+
+let poodr = new Book("Practical Object Oriented Programming in Ruby", "Sandy Mietz");
+
+> Constructor Ran!
+```
+
+### Instance Methods
+
+Instance methods are created like a constructors with their own names.  
+
+```javascript
+const SALESTAX = 0.08;
+
+class Book {
+  constructor(title, author, price) {
+    this.title = title;
+    this.author = author;
+    this.price = price;
+  }
+  totalPrice() {
+    return this.price * (1.0 + SALESTAX);
+  }
+}
+
+var adaBook = new Book("Poodr", "Sandy Meitz", 100);
+console.log(adaBook.totalPrice());
+> 108
+```
+
+### Getter & Setter Methods
+
+Getter methods are methods which are run when a property with the same name is accessed.  Similarly a Setter method is called when there is an attempt to change a property.  
+
+```javascript
+class Book {
+  constructor(title, author) {
+    console.log("Constructor Run!");
+    this._title = title;
+    this._author = author;
+  }
+
+  // getter methods
+  get author() {
+    return this._author;
+  }
+
+  // setter method
+  set author(newAuthor) {
+    if (typeof newAuthor === "string") {
+      this._author = newAuthor;
+    }
+  }
+
+}
+let poodr = new Book("Practical Object Oriented Programming in Ruby", "Sandy Meitz");
+
+// setter method
+poodr.author = "Chris";
+
+console.log(`${poodr.author}`); // getter method
+> Chris
+```
+
+Notice that the properties are all named __\_propertyName__.  This is a typical naming convention for instance variables you want to keep private.  It's a convention that signals an intent to prevent access outside the class as nothing is really private in JavaScript.  Properties cannot have the same names as our getter and setter methods as doing so will result in infinite recursion.  
+
+**Question**:  Why would naming the property the same as the getter or setter method result in infinite recursion?
+
+**Exercise**:  Create getter and setter methods for title.
+
+### Static Methods
+
+Static methods are equivalent to class methods in Ruby.  They are attached to the class rather than instances of the class.
+
+
+```javascript
+class Book {
+  constructor(title, author, price) {
+    this.title = title;
+    this.author = author;
+    this.price = price;
+  }
+
+  static compare(bookA, bookB) {
+    return bookA.price - bookB.price;
```

Comment:
This doesn't really feel like the logic is doing a comparison. Maybe return the title of the book with a higher (or lower) price?